### PR TITLE
feat(testing): use zones to avoid the need for injectAsync

### PR DESCRIPTION
**Pull request by juliemr**
Use a zone counting timeouts and microtasks to determine when a test
is finished, instead of requiring the test writer to use
injectAsync and return a promise.

See #5322

### CODE REVIEW

**Comment by juliemr** [comment]
cc @vicb @IgorMinar 

Zones seem to work smoothly and remove complexity for the user for the async tests. All tests which use `inject` are now being run asynchronously, which as we discussed before adds a very small (couple ms) penalty time to each test. I like this overall - thoughts?

**Comment by juliemr** [comment]
Note that this isn't actually a breaking change, because both `inject` and `injectAsync` do the same thing now and the return value doesn't matter. A cleanup would remove `injectAsync` entirely.

**Comment by PascalPrecht** [comment]
This is awesome Julie!

File: modules/angular2/src/testing/testing.ts [hunk 1]
```diff
@@ -68,37 +69,72 @@ function _isPromiseLike(input): boolean {
   return input && !!(input.then);
 }
 
+function runInTestZone(fnToExecute, finishCallback, failCallback) {
+  var pendingMicrotasks = 0;
+  var pendingTimeouts = [];
+
+  var ngTestZone = (<NgZoneZone>global.zone)
```
Comment (vicb) [inline modules/angular2/src/testing/testing.ts]: `<Zone>global.zone`

File: modules/angular2/src/testing/testing.ts [hunk 2]
```diff
@@ -68,37 +69,72 @@ function _isPromiseLike(input): boolean {
...
+                       .fork({onError: function(e) { failCallback(e); }})
```
Comment (vicb) [inline modules/angular2/src/testing/testing.ts]: could you fork only once here (merging the 2 objects) ?

File: modules/angular2/src/testing/testing.ts (hunk 1 above)
Comment (juliemr) [inline modules/angular2/src/testing/testing.ts]: Good catch, done.

File: modules/angular2/src/testing/testing.ts (hunk 2 above)
Comment (juliemr) [inline modules/angular2/src/testing/testing.ts]: Yes, done.

File: modules/angular2/test/testing/testing_public_spec.ts [hunk 3]
```diff
@@ -182,40 +184,28 @@ export function main() {
...
-    it('should fail when return was forgotten in beforeEach', () => {
-      expect(() => {
-        patchJasmineBeforeEach();
-        beforeEach(injectAsync([], () => { return true; }));
-      })
-          .toThrowError('Error: injectAsync was expected to return a promise, but the ' +
-                        ' returned value was: true');
-      restoreJasmineBeforeEach();
-    });
+    it('should fail when an asyncornous error is thrown', (done) => {
```
Comment (vicb) [inline modules/angular2/test/testing/testing_public_spec.ts]: typo: asynchronous

File: modules/angular2/test/testing/testing_public_spec.ts [hunk 4]
```diff
@@ -157,14 +157,16 @@ export function main() {
...
       originalJasmineIt = jasmine.getEnv().it;
       jasmine.getEnv().it = (description: string, fn) => {
-        var done = () => {};
-        (<any>done).fail = (err) => { throw new Error(err) };
+        var done = () => { deferred.resolve() };
+        (<any>done).fail = (err) => { deferred.reject(err) };
         fn(done);
         return null;
-      }
+      } return deferred.promise;
```
Comment (vicb) [inline modules/angular2/test/testing/testing_public_spec.ts]: missing ";"

**Comment by vicb** [comment]
A few minor comments, otherwise LGTM.
Great PR !

**Comment by IgorMinar** [comment]
@wardbell fyi

**Comment by IgorMinar** [comment]
looks great! Please deprecate injectAsync and add docs about the async behavior for `inject`

**Comment by IgorMinar** [comment]
OMG, I just realized that this might mean that we don't need to refactor all of our tests in `angular2/angular2` to return promises. Now I'm really stoked.

Do we properly detect errors and surface them or will we still see jasmine timeouts when an error occurs in an async test?

**Comment by juliemr** [comment]
Errors should be properly output: As in this test:

```
    it('should fail when an asynchronous error is thrown', (done) => {
      var itPromise = patchJasmineIt();

      it('throws an async error',
         inject([], () => { setTimeout(() => { throw new Error('bar'); }, 0); }));

      itPromise.then(() => { done.fail('Expected function to throw, but it did not'); }, (err) => {
        expect(err.message).toEqual('bar'); // <-- bar not just timeout!
        done();
      });
      restoreJasmineIt();
    });
```

**Comment by wardbell** [comment]
Big win!  Looking forward to trying this.

**Comment by juliemr** [comment]
Deprecation notice and docs are done.

**Comment by IgorMinar** [comment]
@juliemr what if the error is trapped in a promise (e.g. inside of the TestComponentBuilder's promise?)

Also, can you send a follow up PR that deprecates AsyncTestCompleter and turns it into noop?

**Comment by Foxandxss** [comment]
Ohh. Brilliant. That is one of the most needed sugars for NG2. This could potentially fix #5060.

**Comment by juliemr** [comment]
@IgorMinar oh shoot, that is a problem. At the moment, this is passing:

```js
      iit('should fail with an error from a promise', inject([], () => {
        var deferred = PromiseWrapper.completer();
        deferred.promise.then(() => {
          expect(1).toEqual(2);
        });

        deferred.reject('this error');
      }));
```

let me think about it...

**Comment by juliemr** [comment]
FYI: The problem is that the error being thrown for unhandled promise rejection is not in the execution context that the zone is following.

**Comment by juliemr** [comment]
So after some reading, I think that until the promises spec changes, there's nothing we can do to automatically catch those unhandled promise rejections.

We can ask users to return a promise (if applicable) in their test, and handle errors on that for them. I think that's a good step for now, and we still get rid of having `injectAsync`. Thoughts?

**Comment by IgorMinar** [comment]
What do you mean? Zone is following all async events started from within
each it.

**Comment by IgorMinar** [comment]
i think we can keep track of these thing, but it requires some invasive
patching. I added an item to my zone.js whishlist:
https://docs.google.com/document/d/1vvTs3v2jjRn-EIymMGCXWIZb4hJQGxUznbXzjBwvnag/edit

**Comment by IgorMinar** [comment]
For now, let's support return value and make AsyncCompleter a noop and
deprecate it.

**Comment by juliemr** [comment]
Done - returned promise will now cause a failure if it's rejected.

Will do the AsyncCompleter noop in a different PR.

**Comment by juliemr** [comment]
FYI the spec change that will make rejected promise tracking possible w/o patching is approved: Eventually, we'll have the ability to track promise rejection in plain JS: https://github.com/tc39/ecma262/pull/76

**Comment by juliemr** [comment]
issue for AsyncTestCompleter follow-up: https://github.com/angular/angular/issues/5443

**Comment by alxhub** [comment]
Hi Julie - your second commit still says "For squashing:". Can you reword it to follow the style guide?

**Comment by juliemr** [comment]
@alxhub done (I actually just squashed it). Sorry I had parsed your initial instructions incorrectly.

**Comment by juliemr** [comment]
I added the merge label yesterday but it doesn't seem to have stuck :(

This is ready for merge as far as I'm concerned.

**Comment by vsavkin** [comment]
Could you rebase?

**Comment by juliemr** [comment]
@vsavkin done!

**Comment by IgorMinar** [comment]
let's get this in! :)

**Comment by vsavkin** [comment]
Merged via https://github.com/angular/angular/commit/0c9596ae2b31135711fba1d132f8e5d349a1e095
